[SPARK-1406] Mllib pmml model export

selvinsource opened the pull request:
See PDF attached to the JIRA issue 1406.

The contribution is my original work and I license the work to the project under the project's open source license.

selvinsource committed:
export and pmml export traits
kmeans test implementation

selvinsource committed:
added javadoc and export model type in case there is a need to support
other types of export (not just PMML)

selvinsource committed:
added scala suite tests
added saveLocalFile to ModelExport trait

selvinsource committed:
kmeans pmml export implementation

selvinsource committed:
complete suite tests

selvinsource committed:
adjust scala style in order to compile

selvinsource committed:
Update LICENSE

added https://github.com/jpmml/jpmml-model library under BSD-style licenses
the library is used to export the mining model to the PMML format

selvinsource committed:
Merge remote-tracking branch 'origin/master' into mllib_pmml_model_export_SPARK-1406

selvinsource committed:
fixed scala style to run tests

selvinsource committed:
fixed cluster export

selvinsource committed:
removed serialization part, this will be part of the ModelExporter
helper object

selvinsource committed:
set it to private

selvinsource committed:
set it to private

selvinsource committed:
new helper object to serialize the models to pmml format

selvinsource committed:
reordered the import accordingly to the guidelines

selvinsource committed:
updated imports

selvinsource committed:
added pmml version to pmml model
changed the copyright to spark

selvinsource committed:
updated some commented tests to use the new ModelExporter object
reordered the imports

selvinsource committed:
removed version, by default is set to 4.2 (latest from jpmml)
removed copyright

selvinsource committed:
[SPARK-1406] added pmml export for LinearRegressionModel,
RidgeRegressionModel and LassoModel

selvinsource committed:
Merge remote-tracking branch 'upstream/master' into mllib_pmml_model_export_SPARK-1406

selvinsource committed:
[SPARK-1406] Adjusted imported order according to the guidelines

selvinsource committed:
[SPARK-1406] Added target field to the regression model for completeness
Adjusted unit test to deal with this change

selvinsource committed:
[SPARK-1406] Fixed scala style

selvinsource committed:
Merge remote-tracking branch 'upstream/master' into mllib_pmml_model_export_SPARK-1406

selvinsource committed:
[SPARK-1406] added linear SVM PMML export

selvinsource committed:
added logistic regression

selvinsource committed:
[SPARK-1406] Adjusted logistic regression export description and target
categories

jkbradley commented:
@selvinsource  This will be great to get PMML support!  I'll try to give some feedback soon.

Quick question: Do you know if there is a freely available PMML validator we could use for checking syntax?  Or is JPMML checked thoroughly already?  (The AWS and iGoogle links for the Zementis validator seem to be broken now.)

selvinsource commented:
@jkbradley 
I use JPMML to verify the exported model produces the same results, here the details of my tests:
https://github.com/selvinsource/spark-pmml-exporter-validator

mengxr commented:
@srowen @selvinsource From user perspective, this is the code they need to export a model to PMML:

~~~
import org.apache.spark.mllib.export.ModelExporter

val kmeansModel = ...
ModelExporter.toPMML(kmeansModel, "/path")
~~~

They need to remember two things: 1) `ModelExporter`'s name and its package, 2) whether a model is exportable to PMML or not because `toPMML` takes `Any`. As a user, I would prefer the following

~~~
val kmeansModel = ...
val pmml = kmeansModel.toPMML
~~~

We can add a trait called `PMMLExportable`, which contains `toPMML`. (We need to discuss the return type of `toPMML`.) So users know whether a model can be exported to PMML or not. For the implementation, I agree with @srowen that `.pmml` package should be sufficient because there is no other exchangeable formats for predictive models. The real export code for each model should live under this package as private and `toPMML` in each model becomes a simple wrapper.

Due to lacking feature attributes, I'd like to mark everything as private for now until we add ML attributes.

selvinsource commented:
Either approach has its pros and cons and I don't have a strong argument in favor of any of the two.

R, for instance, use a similar approach to what I have done so far and suggested by @srowen
    library(pmml)
    model0 <- lm(Sepal.Length~., data=iris[,-5])
    model <- pmml(model0)
but I understand @mengxr  point of view, a user would prefer to just call `.toPMML` on the model.

If we all agree this is the way forward, I can re-organize the code as suggested by @mengxr having the supported models extending the PMMLExportable trait.

mengxr commented:
The R case is slightly different because the `pmml` package is maintained separately from other packages. @srowen , are you okay with the interface I proposed?

srowen commented:
@mengxr I like keeping the export together in one package. Since I imagine there will likely only ever be 1 or 2 model export formats (really, I imagine 1) I think using a trait is OK. That is, I don't feel too bad about treating PMML fairly specially. 

The `PMML` abstraction from JPMML is pretty solid and could be the return type of such a `toPMML` API method, although I hesitate to bake in a third-party library class to the public API. I suppose the return type could just be `String`, containing an XML document, or at best a generic object representing the XML DOM?

Yes, definitely mark things private / developer API until this stabilizes.

Writing PMML serializations is a moderate PITA, but using the JPMML library internally makes it not so bad. I have some serializations already written I could chuck into the pot once the framework is in place (decision forest, k-means clustering model)

jkbradley commented:
@srowen  I expect us to have 2 model formats (PMML, plus a Spark-specific one supporting distributed models).  +1 for returning a String or some type which lets us hide the JPMML dependency.

mengxr commented:
@srowen Yes, the implementation would live under the `mllib.pmml` package and we add `toPMML` to models that we support. For the return type, exposing third-party APIs is not recommended in Spark. We used Guava's Optional, which turned out to be a bad choice for maintenance. There are pros/cons for other types.

1. String. Easy to save/load, but it may not be able to handle super large XML.
2. `Result` from `javax.xml`. Officially supported by Java, but most users don't know how to use it.

So it really depends on how big the model is and whether it is necessary to output XML as a stream.

selvinsource committed:
Merge remote-tracking branch 'upstream/master' into mllib_pmml_model_export_SPARK-1406

selvinsource committed:
[SPARK-1406] Added a PMMLExportable interface
Restructured code in a new package mllib.pmml
Supported models implements the new PMMLExportable interface:
LogisticRegression, SVM, KMeansModel, LinearRegression, RidgeRegression,
Lasso

selvinsource committed:
[SPARK-1406] Added PMMLExportable to supported models

SparkQA commented:
[Test build #27046 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27046/consoleFull) for   PR 3062 at commit [`f46c75c`](https://github.com/apache/spark/commit/f46c75cdbea8170c332026ffd05d5ca1667d50c6).
 * This patch merges cleanly.

SparkQA commented:
[Test build #27046 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27046/consoleFull) for   PR 3062 at commit [`f46c75c`](https://github.com/apache/spark/commit/f46c75cdbea8170c332026ffd05d5ca1667d50c6).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (val clusterCenters: Array[Vector]) extends Serializable with PMMLExportable `
  * `trait PMMLExportable `

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27046/
Test FAILed.

selvinsource committed:
[SPARK-1406] Fixed scala style

SparkQA commented:
[Test build #27048 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27048/consoleFull) for   PR 3062 at commit [`7a949d0`](https://github.com/apache/spark/commit/7a949d0b3f5b397f4df4fcd70f3658cc66e8f04e).
 * This patch merges cleanly.

selvinsource commented:
@mengxr @srowen 
I created the PMMLExportable trait and moved all the code under the package mllib.pmml.
Supported models implements now this trait and the usage is as you envisaged:
`val pmml = kmeansModel.toPMML()`

I added also extra export options, respectively File and OutputStream:
`kmeansModel.toPMML("/tmp/kmeans.xml")`
`kmeansModel.toPMML(System.out)`
Feel free to review those, I would suggest to keep at least the OutputStream which is well known.

SparkQA commented:
[Test build #27048 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/27048/consoleFull) for   PR 3062 at commit [`7a949d0`](https://github.com/apache/spark/commit/7a949d0b3f5b397f4df4fcd70f3658cc66e8f04e).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (val clusterCenters: Array[Vector]) extends Serializable with PMMLExportable `
  * `trait PMMLExportable `

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/27048/
Test PASSed.

jkbradley commented:
@selvinsource I like the output options which write to disk, but what about supporting writing to distributed file systems?  It's more common in Spark to use URIs instead of local paths, but it might be reasonable to make an exception for PMML since it might often be used for small models.  What are your thoughts?

selvinsource commented:
@jkbradley I could add that too in addition or alternative to the local file.

In terms of implementation, I was thinking that the quick way of doing it would be:

    def toPMML(sc: SparkContext, path: String) : Unit = {
       val pmml = toPMML() //this export the model to a String
       sc.parallelize(Array(pmml)).saveAsTextFile(path)
   }

Is this OK or would you suggest a different approach?

jkbradley commented:
I see, so the SparkContext makes it clear it's being saved to a DFS.  That sounds great to me.  (Would you want to enforce numPartitions = 1 so that it is stored on a single partition?)

selvinsource commented:
Yeah that makes sense, I don't think the xml would be that big to require multiple partitions:

    sc.parallelize(Array(pmml),1).saveAsTextFile(path)

jkbradley commented:
Sounds good!

selvinsource committed:
[SPARK-1406] Added export of pmml to distributed file system using the
spark context

SparkQA commented:
[Test build #28149 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28149/consoleFull) for   PR 3062 at commit [`b25bbf7`](https://github.com/apache/spark/commit/b25bbf7ae7eb92541f59f5cfe7495280d60fca43).
 * This patch merges cleanly.

SparkQA commented:
[Test build #28149 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/28149/consoleFull) for   PR 3062 at commit [`b25bbf7`](https://github.com/apache/spark/commit/b25bbf7ae7eb92541f59f5cfe7495280d60fca43).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/28149/
Test PASSed.

damiencarol commented:
Any news on this one?

mengxr commented:
I'm going to make a pass this week. The target is 1.4.

selvinsource committed:
Merge remote-tracking branch 'upstream/master' into
mllib_pmml_model_export_SPARK-1406

Conflicts:
	mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala

SparkQA commented:
[Test build #30562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30562/consoleFull) for   PR 3062 at commit [`b8823b0`](https://github.com/apache/spark/commit/b8823b022129ad49836b90745a8aa397387c60cf).

selvinsource commented:
@mengxr I aligned the branch with the master and resolved a conflict so that it should be easier for you to review.

SparkQA commented:
[Test build #30562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30562/consoleFull) for   PR 3062 at commit [`b8823b0`](https://github.com/apache/spark/commit/b8823b022129ad49836b90745a8aa397387c60cf).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (val clusterCenters: Array[Vector]) extends Saveable with Serializable with PMMLExportable `
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `FastInfoset-1.2.12.jar`
   * `istack-commons-runtime-2.16.jar`
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `jsr173_api-1.0.jar`
   * `pmml-model-1.1.7.jar`
   * `pmml-schema-1.1.7.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30562/
Test FAILed.

selvinsource committed:
fixed scala style

selvinsource committed:
fixed scala issue

SparkQA commented:
[Test build #30564 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30564/consoleFull) for   PR 3062 at commit [`1676e15`](https://github.com/apache/spark/commit/1676e15ff5bec346466a095a593efdd324f6b515).

SparkQA commented:
[Test build #30564 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30564/consoleFull) for   PR 3062 at commit [`1676e15`](https://github.com/apache/spark/commit/1676e15ff5bec346466a095a593efdd324f6b515).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `FastInfoset-1.2.12.jar`
   * `istack-commons-runtime-2.16.jar`
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `jsr173_api-1.0.jar`
   * `pmml-model-1.1.7.jar`
   * `pmml-schema-1.1.7.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30564/
Test PASSed.

mengxr committed:
fix code style

mengxr commented:
@selvinsource Could you check whether the following libraries have Apache-compatible license?

* FastInfoset-1.2.12.jar
* istack-commons-runtime-2.16.jar
* jsr173_api-1.0.jar
* pmml-model-1.1.7.jar
* pmml-schema-1.1.7.jar

This PR also upgrades `jaxb` and `activation`. We need to make sure this is a compatible change. Or we can use an older version of `jpmml`.

I will send you a PR that updates the code style. Then I will make another pass on the implementation.

mengxr commented:
@selvinsource I sent you a PR at https://github.com/selvinsource/spark/pull/1 to update the code style.

selvinsource committed:
Merge pull request #1 from mengxr/SPARK-1406

Update code style

SparkQA commented:
[Test build #30646 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30646/consoleFull) for   PR 3062 at commit [`e2313df`](https://github.com/apache/spark/commit/e2313df33560bfd1e8377ff33be8a67d5ad0922a).

selvinsource commented:
@mengxr I merged your PR, I will review the libraries and let you know.

mengxr commented:
@selvinsource Thanks! There are some indentation issues in my PR. Could you help fix it?

The latest version of `jpmml-model` requires Java 7 or newer, while Spark needs to support Java 6, 7, and 8. @vruusmann Do you know what is the newest version of `jpmml-model` that supports Java 6?

mengxr committed:
more code style

vruusmann commented:
@mengxr Branches 1.0.X and 1.1.X are Java 6. The new branch 1.2.X is Java 7.

The latest Java 6 compatible version is 1.1.15.

SparkQA commented:
[Test build #30646 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30646/consoleFull) for   PR 3062 at commit [`e2313df`](https://github.com/apache/spark/commit/e2313df33560bfd1e8377ff33be8a67d5ad0922a).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `FastInfoset-1.2.12.jar`
   * `istack-commons-runtime-2.16.jar`
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `jsr173_api-1.0.jar`
   * `pmml-model-1.1.7.jar`
   * `pmml-schema-1.1.7.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30646/
Test PASSed.

selvinsource committed:
Merge pull request #2 from mengxr/SPARK-1406

more code style

SparkQA commented:
[Test build #30661 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30661/consoleFull) for   PR 3062 at commit [`a0a55f7`](https://github.com/apache/spark/commit/a0a55f7add73b0fbbb27b285e9316cb23c716aaf).

SparkQA commented:
[Test build #30661 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30661/consoleFull) for   PR 3062 at commit [`a0a55f7`](https://github.com/apache/spark/commit/a0a55f7add73b0fbbb27b285e9316cb23c716aaf).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `FastInfoset-1.2.12.jar`
   * `istack-commons-runtime-2.16.jar`
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `jsr173_api-1.0.jar`
   * `pmml-model-1.1.7.jar`
   * `pmml-schema-1.1.7.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30661/
Test PASSed.

selvinsource commented:
@mengxr

FastInfoset-1.2.12.jar -> http://en.wikipedia.org/wiki/Fast_Infoset -> Apache License 2.0
istack-commons-runtime-2.16.jar -> https://istack-commons.java.net/ -> CDDL v1.1
jaxb* -> https://jaxb.java.net/ -> CDDL v1.1
pmml* -> https://github.com/jpmml/jpmml-model -> BSD 3-Clause License

Accordingly to http://www.apache.org/legal/resolved.html CDDL is possible as binary, but please verify too that this is the case.

@vruusmann 
I assume that if JPMML-Model is licensed under the BSD 3-Clause License this is compatible with CDDL v1.1?

selvinsource commented:
@mengxr @vruusmann 
I am using 1.1.7, therefore compatible with Java 6. Should I update it to the latest Java 6 supported version 1.1.15? Any benefit in doing so?

srowen commented:
@selvinsource have a look at our `NOTICE` file to see how to cite these components with their licenses, and `LICENSE` to see what to do with AL2 deps with their own `NOTICE` file.

vruusmann commented:
@selvinsource The benefits of upgrading the `pmml-model` dependency are more obvious if you are in the business of consuming PMML documents (eg. speed and memory usage improvements). But I still recommend you to make this move.

The upgrade from 1.1.7 to 1.1.13 should be very easy. The upgrade from 1.1.13 to 1.1.14 may require touching Java source code here and there (eg. rearranging the arguments of value constructors, correcting the case of selected method names).

The upgrade will change the list of your transitive dependencies. So, you should probably to the upgrade first, and edit `LICENSE` and `NOTICE` files second.

Moreover, I think that after the upgrade it will be safe to exclude `FastInfoset` and `istack-commons-runtime`  dependencies from the build (using Maven's dependency exclusion mechanism) without affecting the PMML exporter functionality. These dependencies take care of managing SAX Locator information when unmarshalling. But since you are creating new PMML documents in memory from scratch, then they are not needed.

srowen commented:
+1 to updating to the latest 1.1.x JPMML. I have had good experiences with it.

mengxr commented:
@selvinsource Let's upgrade to 1.1.15 and remove FastInfoset and istack-commons-runtime from the build as @vruusmann recommended. Then we need to check the transitive dependencies again.

Is JAXB backward compatible?

selvinsource committed:
[SPARK-1406] Updated pmml model lib to 1.1.15, latest Java 6 compatible

selvinsource committed:
[SPARK-1406] Exclude transitive dependency for pmml model

selvinsource commented:
@mengxr  @vruusmann, I tried the upgrade to 1.1.15 and it requires some code change, I will look into that.

Once we have the new list of dependencies I will review NOTICE and LICENSE as suggested by @srowen.

selvinsource committed:
[SPARK-1406] Update code to latest pmml model

mengxr commented:
@selvinsource Sounds good. Please keep the code style consistent:)

SparkQA commented:
[Test build #30733 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30733/consoleFull) for   PR 3062 at commit [`25dce33`](https://github.com/apache/spark/commit/25dce336065e39da8492c6d8379a21ab3733cd45).

SparkQA commented:
[Test build #30733 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/30733/consoleFull) for   PR 3062 at commit [`25dce33`](https://github.com/apache/spark/commit/25dce336065e39da8492c6d8379a21ab3733cd45).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/30733/
Test PASSed.

vruusmann commented:
@selvinsource @mengxr I created an small project https://github.com/vruusmann/jpmml-test in order to demonstrate how it's possible to reduce the number of dependencies.

The idea is that you should keep JAXB dependencies (scope: `compile`) when building for Java 1.6, but you can safely  exclude them (scope: `provided`) when building for Java 1.7 or newer. 

Here's the output of `mvn dependency:tree` command for Java 1.6 (the size of the resulting uber-JAR file is 1.8 MB):
```
[INFO] org.jpmml:pmml-test:jar:1.0-SNAPSHOT
[INFO] +- org.jpmml:pmml-model:jar:1.1.15:compile
[INFO] |  +- org.jpmml:pmml-agent:jar:1.1.15:compile
[INFO] |  \- org.jpmml:pmml-schema:jar:1.1.15:compile
[INFO] \- com.sun.xml.bind:jaxb-impl:jar:2.2.7:compile
[INFO]    \- com.sun.xml.bind:jaxb-core:jar:2.2.7:compile
[INFO]       \- javax.xml.bind:jaxb-api:jar:2.2.7:compile
```

Here's the same for Java 1.7 (the size of the uber-JAR file has shrunk by 1.2 MB to 0.6 MB; all the remaining dependencies `pmml-model`, `pmml-schema` and `pmml-agent` are covered by the BSD 3-Clause License):
```
[INFO] org.jpmml:pmml-test:jar:1.0-SNAPSHOT
[INFO] +- org.jpmml:pmml-model:jar:1.1.15:compile
[INFO] |  +- org.jpmml:pmml-agent:jar:1.1.15:compile
[INFO] |  \- org.jpmml:pmml-schema:jar:1.1.15:compile
[INFO] \- com.sun.xml.bind:jaxb-impl:jar:2.2.7:provided
[INFO]    +- com.sun.xml.bind:jaxb-core:jar:2.2.7:provided
[INFO]    |  +- javax.xml.bind:jaxb-api:jar:2.2.7:provided
[INFO]    |  \- com.sun.istack:istack-commons-runtime:jar:2.16:provided
[INFO]    \- com.sun.xml.fastinfoset:FastInfoset:jar:1.2.12:provided
[INFO]       \- javax.xml.bind:jsr173_api:jar:1.0:provided
```

JAXB dependencies were made part of Java runtime sometimes in the middle of Java 1.6 (for example, 1.6_u01 does not have them, whereas 1.6_u45 does). So, it's probably possible to fine-tune the activation of the Maven build profile some more.

selvinsource commented:
@mengxr code updated, style should be consistent :)!

@vruusmann thanks for your test project, I did the same last night (mvn dependency:tree) to find out what were the transitive dependencies and I explicitely excluded istack-commons-runtime and FastInfoset as you suggested. This leave us with JAXB which I am going to review in terms of license (still need to check the NOTICE and LICENSE files).

selvinsource commented:
@srowen 
https://github.com/apache/spark/blob/master/NOTICE already has a reference to JAXB:
     (CDDL 1.1) (GPL2 w/ CPE) JAXB API bundle for GlassFish V3 (javax.xml.bind:jaxb-api:2.2.2 - https://jaxb.dev.java.net/)
     (CDDL 1.1) (GPL2 w/ CPE) JAXB RI (com.sun.xml.bind:jaxb-impl:2.2.3-1 - http://jaxb.java.net/)

Is it just a matter of updating the NOTICE file with the new version 2.2.7?

srowen commented:
@selvinsource it's not necessary to update the version. I think it's there since I auto-generated these lists from a Maven report at the outset.

selvinsource commented:
@srowen that sounds good :+1:

mengxr commented:
@selvinsource I tested KMeans.toPMML, which looks good. Linear regressors look good too, but for linear classifiers `LogisticRegressionModel` and `SVMModel`:

1. We should use `functionName="classification"` (as documented in http://www.dmg.org/v4-2-1/Regression.html). The following is an example logistic regression model in PMML (copied from the url above):
   ~~~xml
<PMML xmlns="http://www.dmg.org/PMML-4_2" version="4.2">
  <Header copyright="DMG.org"/>
  <DataDictionary numberOfFields="5">
    <DataField name="age" optype="continuous" dataType="double"/>
    <DataField name="work" optype="continuous" dataType="double"/>
    <DataField name="sex" optype="categorical" dataType="string">
      <Value value="0"/>
      <Value value="1"/>
    </DataField>
    <DataField name="minority" optype="categorical" dataType="integer">
      <Value value="0"/>
      <Value value="1"/>
    </DataField>
    <DataField name="jobcat" optype="categorical" dataType="string">
      <Value value="clerical"/>
      <Value value="professional"/>
      <Value value="trainee"/>
      <Value value="skilled"/>
    </DataField>
  </DataDictionary>
  <RegressionModel modelName="Sample for logistic regression" functionName="classification" algorithmName="logisticRegression" normalizationMethod="softmax" targetFieldName="jobcat">
    <MiningSchema>
      <MiningField name="age"/>
      <MiningField name="work"/>
      <MiningField name="sex"/>
      <MiningField name="minority"/>
      <MiningField name="jobcat" usageType="target"/>
    </MiningSchema>
    <RegressionTable intercept="46.418" targetCategory="clerical">
      <NumericPredictor name="age" exponent="1" coefficient="-0.132"/>
      <NumericPredictor name="work" exponent="1" coefficient="7.867E-02"/>
      <CategoricalPredictor name="sex" value="0" coefficient="-20.525"/>
      <CategoricalPredictor name="sex" value="1" coefficient="0.5"/>
      <CategoricalPredictor name="minority" value="0" coefficient="-19.054"/>
      <CategoricalPredictor name="minority" value="1" coefficient="0"/>
    </RegressionTable>
    <RegressionTable intercept="51.169" targetCategory="professional">
      <NumericPredictor name="age" exponent="1" coefficient="-0.302"/>
      <NumericPredictor name="work" exponent="1" coefficient="0.155"/>
      <CategoricalPredictor name="sex" value="0" coefficient="-21.389"/>
      <CategoricalPredictor name="sex" value="1" coefficient="0.1"/>
      <CategoricalPredictor name="minority" value="0" coefficient="-18.443"/>
      <CategoricalPredictor name="minority" value="1" coefficient="0"/>
    </RegressionTable>
    <RegressionTable intercept="25.478" targetCategory="trainee">
      <NumericPredictor name="age" exponent="1" coefficient="-0.154"/>
      <NumericPredictor name="work" exponent="1" coefficient="0.266"/>
      <CategoricalPredictor name="sex" value="0" coefficient="-2.639"/>
      <CategoricalPredictor name="sex" value="1" coefficient="0.8"/>
      <CategoricalPredictor name="minority" value="0" coefficient="-19.821"/>
      <CategoricalPredictor name="minority" value="1" coefficient="0.2"/>
    </RegressionTable>
    <RegressionTable intercept="0.0" targetCategory="skilled"/>
  </RegressionModel>
</PMML>
~~~

2. `threshold` is not encoded in the PMML, which should be there.

3. We don't support multinomial logistic regression model. So we should check the number of classes and throw an exception if the model is multiclass.

I think we should fix 1&2 and throw exception for 3 in this PR. The multi-class support can come later. Since the feature freeze deadline is close, do you have time to update this PR before next Wed (leaving some buffer for review)? Thanks!

selvinsource committed:
[SPARK-1406] Throw IllegalArgumentException when exporting a multinomial
logistic regression

selvinsource commented:
@mengxr 
3)
I have done point 3, if it is a multinomial logistic regression the export will cause an IllegalArgumentException. I didn't realize now Logistic supports multiple classes. As you said we can add support for this in a new PR.

1)
For `LogisticRegressionModel` I do use `functionName =  "classification"`, see an example of an exported model
https://github.com/selvinsource/spark-pmml-exporter-validator/blob/master/src/main/resources/exported_pmml_models/logisticregression.xml 
This can be loaded using JPMML as `ClassificationMap`, see
https://github.com/selvinsource/spark-pmml-exporter-validator/blob/master/src/main/java/org/selvinsource/spark_pmml_exporter_validator/SparkPMMLExporterValidator.java

For `SVMModel`, I did look into exporting it as a PMML classification `RegressionModel` like I did with the `LogisticRegressionModel`, however there isn't a PMML `normalizationMethod` that I can use, they all convert the prediction into probability and `SVMModel` simply compare a scalar to a threshold.
Because of this I added a comment in the exported regression model: "linear SVM: if predicted value > 0, the outcome is positive, or negative otherwise".

2)
I don't see PMML `RegressionModel` supporting a threshold, @vruusmann can you confirm this is the case, or am I missing something :)?
For `SVMModel` export I could add it to the comment, now it is hard coded to 0 but could be "linear SVM: if predicted value > `threshold variable` the outcome is positive, or negative otherwise".

mengxr commented:
Sorry, I only verified SVM and thought logistic regression was implemented the same way. For SVM, could we try `normalizationMethod = none` and set the threshold as the intercept of the second regression table? 

~~~xml
 <RegressionModel functionName="classification"  normalizationMethod="none" ...
    <RegressionTable targetCategory="YES" ...
    <RegressionTable targetCategory="NO" intercept="0.05"
    ...
~~~

Maybe we can do the same for binary logistic regression. I don't know whether this is the correct way to do it by looking at the PMML specification. Need help from @vruusmann

selvinsource commented:
@mengxr for SVM, I manually tried what you suggested and it looks good.

I loaded the example below in JPMML and evaluated it as Classification map, indeed the intercept on the NO category acts as threshold when `normalizationMethod = none`.
Here the example:
<code>
<?xml version="1.0" encoding="UTF-8" standalone="yes"?>
<PMML xmlns="http://www.dmg.org/PMML-4_2">
    <Header description="linear SVM: if predicted value &gt; 0, the outcome is positive, or negative otherwise">
        <Application name="Apache Spark MLlib" version="1.4.0-SNAPSHOT"/>
        <Timestamp>2015-04-27T06:58:22</Timestamp>
    </Header>
    <DataDictionary numberOfFields="10">
        <DataField name="field_0" optype="continuous" dataType="double"/>
        <DataField name="field_1" optype="continuous" dataType="double"/>
        <DataField name="field_2" optype="continuous" dataType="double"/>
        <DataField name="field_3" optype="continuous" dataType="double"/>
        <DataField name="field_4" optype="continuous" dataType="double"/>
        <DataField name="field_5" optype="continuous" dataType="double"/>
        <DataField name="field_6" optype="continuous" dataType="double"/>
        <DataField name="field_7" optype="continuous" dataType="double"/>
        <DataField name="field_8" optype="continuous" dataType="double"/>
        <DataField name="target" optype="categorical" dataType="string"/>
    </DataDictionary>
    <RegressionModel modelName="linear SVM: if predicted value &gt; 0, the outcome is positive, or negative otherwise" functionName="classification" normalizationMethod="none">
        <MiningSchema>
            <MiningField name="field_0" usageType="active"/>
            <MiningField name="field_1" usageType="active"/>
            <MiningField name="field_2" usageType="active"/>
            <MiningField name="field_3" usageType="active"/>
            <MiningField name="field_4" usageType="active"/>
            <MiningField name="field_5" usageType="active"/>
            <MiningField name="field_6" usageType="active"/>
            <MiningField name="field_7" usageType="active"/>
            <MiningField name="field_8" usageType="active"/>
            <MiningField name="target" usageType="target"/>
        </MiningSchema>
        <RegressionTable intercept="-1.2973802920137774" targetCategory="1">
            <NumericPredictor name="field_0" coefficient="-0.0818303650185629"/>
            <NumericPredictor name="field_1" coefficient="0.5609579878511747"/>
            <NumericPredictor name="field_2" coefficient="0.1382792114252377"/>
            <NumericPredictor name="field_3" coefficient="0.07497131265977852"/>
            <NumericPredictor name="field_4" coefficient="-0.47760356523751296"/>
            <NumericPredictor name="field_5" coefficient="0.3817837986572615"/>
            <NumericPredictor name="field_6" coefficient="-0.23753782335208481"/>
            <NumericPredictor name="field_7" coefficient="0.2548602390316011"/>
            <NumericPredictor name="field_8" coefficient="-0.10271528637619945"/>
        </RegressionTable>
        <RegressionTable intercept="0.0" targetCategory="0"/>
    </RegressionModel>
</PMML>
</code>

However, I noticed that if the SVM model threshold is set to None, it simply displays the margin (which is how it is implemented now in the pmml exporter). 
My question is, should we support both? If `threshold = None`, export as regression (like it is implemented now), if `threshold <> None`, export as binary classification (as you suggested). What do you think?

selvinsource commented:
For binary logistic regression, using the same principle (intercept as threshold), doing some maths, we could set:
`intercept = -ln(1/threshold - 1)`

selvinsource commented:
Here my thinking.

When normalizationMethod = logit, the predicted value is computed as 
`pj = 1 / ( 1 + exp( -yj ) )`
When we set `intercept = 0` for `targetCategory=0` => `yj = 0` => `pj = 1 / ( 1 + exp( 0 ) ) = 0.5`, which is in fact the default threshold for binary logistic classification.

Say we want `pj = threshold` (whereas threshold could be any value between 0 and 1, different to the default 0.5), we need to compute yj in terms of pj (which is given)
`yj = -ln(1/pj-1)` where yj is the intercept and pj is the threshold, hence `intercept = -ln(1/threshold - 1)`.

Need to be tested, meantime, what do @vruusmann @mengxr think?

A pratical example, `model.threshold = 0.6` => `intercept = -ln(1/threshold - 1) = 0.41` (calculated from pmml exporter) => `pj = 1 / ( 1 + exp( -0.41 ) ) = 0.6` (computed by JPMML back to the original threshold).

AmplabJenkins commented:
Can one of the admins verify this patch?

mengxr commented:
@selvinsource Thanks for testing! If the threshold is set to none, I think we should use 0.5 for logistic regression and 0.0 for SVM, maybe with warning log messages. Your proposal on how to set the intercept for NO in logistic regression looks good to me. Please remember to handle the extreme case (when threshold = 1.0 or 0.0).

selvinsource committed:
[SPARK-1406] Binary classification for SVM and Logistic Regression

selvinsource committed:
[SPARK-1406] Fixed extreme cases for logit

selvinsource commented:
@mengxr please review, it should work as expected now.

mengxr commented:
test this please

SparkQA commented:
[Test build #31147 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31147/consoleFull) for   PR 3062 at commit [`30165c4`](https://github.com/apache/spark/commit/30165c4caaea77b5b91ce2bef0ca94cb0454f527).

SparkQA commented:
[Test build #31147 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31147/consoleFull) for   PR 3062 at commit [`30165c4`](https://github.com/apache/spark/commit/30165c4caaea77b5b91ce2bef0ca94cb0454f527).
 * This patch **fails Scala style tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

mengxr commented:
@selvinsource You can run `dev/lint-scala` on your local to check Scala style.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/BinaryClassificationPMMLModelExport.scala:
```diff
@@ -0,0 +1,89 @@
+       var interceptNO = threshold
+       if (RegressionNormalizationMethodType.LOGIT == normalizationMethod) {
+         if (threshold <= 0)
+           interceptNO = -1000
```
Should be `Double.MinValue`.

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/BinaryClassificationPMMLModelExport.scala:
```diff
@@ -0,0 +1,89 @@
+         if (threshold <= 0)
+           interceptNO = -1000
+         else if (threshold >= 1)
+           interceptNO = 1000
```
`Double.MaxValue`

mengxr commented on mllib/src/main/scala/org/apache/spark/mllib/pmml/export/BinaryClassificationPMMLModelExport.scala:
```diff
@@ -0,0 +1,89 @@
+         else if (threshold >= 1)
+           interceptNO = 1000
+         else
+           interceptNO = -math.log(1/threshold -1)
```
space around `/` and after `-`: `math.log(1 / threshold - 1)`

selvinsource committed:
[SPARK-1406] Added Double Min and Max
Fixed scala style

selvinsource committed:
[SPARK-1406] Update JPMML version to 1.1.15 in LICENSE file

selvinsource commented:
@mengxr updated based on your latest comments.

mengxr commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31313 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31313/consoleFull) for   PR 3062 at commit [`852aac6`](https://github.com/apache/spark/commit/852aac6149d4d0c932ff34e959d148419b369983).

SparkQA commented:
[Test build #31313 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31313/consoleFull) for   PR 3062 at commit [`852aac6`](https://github.com/apache/spark/commit/852aac6149d4d0c932ff34e959d148419b369983).
 * This patch **fails MiMa tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31313/
Test FAILed.

mengxr commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31339 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31339/consoleFull) for   PR 3062 at commit [`852aac6`](https://github.com/apache/spark/commit/852aac6149d4d0c932ff34e959d148419b369983).

SparkQA commented:
[Test build #31339 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31339/consoleFull) for   PR 3062 at commit [`852aac6`](https://github.com/apache/spark/commit/852aac6149d4d0c932ff34e959d148419b369983).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`
   * `spark-unsafe_2.10-1.4.0-SNAPSHOT.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31339/
Test FAILed.

mengxr commented:
test this please

AmplabJenkins commented:
Merged build triggered.

AmplabJenkins commented:
Merged build started.

SparkQA commented:
[Test build #31357 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31357/consoleFull) for   PR 3062 at commit [`852aac6`](https://github.com/apache/spark/commit/852aac6149d4d0c932ff34e959d148419b369983).

SparkQA commented:
[Test build #31357 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31357/consoleFull) for   PR 3062 at commit [`852aac6`](https://github.com/apache/spark/commit/852aac6149d4d0c932ff34e959d148419b369983).
 * This patch **passes all tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `class KMeansModel (`
  * `trait PMMLExportable `

 * This patch **adds the following new dependencies:**
   * `jaxb-api-2.2.7.jar`
   * `jaxb-core-2.2.7.jar`
   * `jaxb-impl-2.2.7.jar`
   * `pmml-agent-1.1.15.jar`
   * `pmml-model-1.1.15.jar`
   * `pmml-schema-1.1.15.jar`

 * This patch **removes the following dependencies:**
   * `activation-1.1.jar`
   * `jaxb-api-2.2.2.jar`
   * `jaxb-impl-2.2.3-1.jar`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/31357/
Test PASSed.

mengxr commented:
LGTM. Merged into master. Thanks! @selvinsource I created SPARK-7272 for the user guide and assigned it to you. Let me know if you don't have time to do it. Btw, please submit your PMML evaluator code as a Spark package at spark-packages.org. Due to license issues, we cannot include them as example code in the Spark codebase. But PMML model scoring is quite important.

selvinsource commented:
@mengxr thanks for merging this PR!

I will look into submitting the evaluator at spark-packages if you think it will be useful, I used it as side project to make sure the exported models were working correctly.

On the documentation, please point me in the right direction and I will be happy to help. We can discuss this on SPARK-7272.

selvinsource commented:
@mengxr jpmml evaluator submitted to http://spark-packages.org

tvmanikandan commented:
when PMML support for multi class expected to be released ? I am eagerly waiting PMML multiclass support for Logistic Regression , stochastic gradient  descent (SGD) and SVM.
Please let me know.

-Mani

selvinsource commented:
@tvmanikandan 
we are adding decision tree PMML support for multiclass classification, see https://issues.apache.org/jira/browse/SPARK-8542.

Linear SVMs supports only binary classification, while logistic regression supports both binary and multiclass classification problems. If you would like logistic regression multiclass classification PMML export and the decision tree is not enough, please create a Jira issue under the https://issues.apache.org/jira/browse/SPARK-8545 PMML improvement umbrella.

manugarri commented:
im not sure if this is the right place to ask, but is there any plan to implement PMML export from pyspark? Cant find anything on the pyspark docs.

vruusmann commented:
@manugarri You can export fitted pipeline models to PMML using the [JPMML-SparkML-Package](https://github.com/jpmml/jpmml-sparkml-package) Apache Spark Package. There's a worked-out PySpark example in its README.md file.
